File: library/src/main/java/org/metalscraps/eso/lang/lib/bean/PO.java
package org.metalscraps.eso.lang.lib.bean;

import lombok.AccessLevel;
import lombok.Data;
import lombok.Getter;
import lombok.Setter;
import org.metalscraps.eso.lang.lib.config.FileNames;

import java.util.Comparator;
import java.util.Objects;

/**
 * Created by 안병길 on 2018-01-18.
 * Whya5448@gmail.com
 */

@Data
public class PO implements Comparable<PO> {

    public static Comparator<PO> comparator = (o1, o2) -> {
        if(!Objects.equals(o1.getId1(), o2.getId1())) return o1.getId1() - o2.getId1();
        if(!Objects.equals(o1.getId2(), o2.getId2())) return o1.getId2() - o2.getId2();
        return o1.getId3() - o2.getId3();
    };

	public enum POWrapType {
		WRAP_ALL,
		WRAP_SOURCE,
		WRAP_TARGET
	}

	public PO(String id, String source, String target) {
		source = source.replaceAll("\"\n\"", "");
		target = target.replaceAll("\"\n\"", "");

		this.id = id;
		this.source = source;
		this.target = target;

		String[] ids = id.split("-");
		id1 = Integer.parseInt(ids[0]);
		id2 = Integer.parseInt(ids[1]);
		id3 = Integer.parseInt(ids[2]);

		if(target.equals("")) this.target = source;
		else if(source.equals("")) this.source = target;
	}

	private Integer id1, id2, id3;

	@Setter(AccessLevel.PUBLIC) @Getter(AccessLevel.PUBLIC)
	private String id, source, target, stringFileName;
	private boolean fuzzy = false;
	private FileNames fileName;


	public boolean modifyDoubleQuart(){
		this.source  =source.replace("\"\"", "\"");
		this.target  =target.replace("\"\"", "\"");
		return true;
	}

	public PO wrap(String prefix, String suffix, POWrapType wrapType) {

		if (prefix == null) prefix = "";
		if (suffix == null) suffix = "";

		if(wrapType == POWrapType.WRAP_ALL) {
			if(!source.equals("")) source = prefix + source + suffix;
			if(!target.equals("")) target = prefix + target + suffix;
		} else if(wrapType == POWrapType.WRAP_SOURCE && !source.equals("")) source = prefix + source + suffix;
		else if(wrapType == POWrapType.WRAP_TARGET && !target.equals("")) target = prefix + target + suffix;

		return this;
	}

	public String toStringDefault() { return toCSV(new ToCSVConfig()); }

	public String toCSV(ToCSVConfig toCSVConfig) {
		String translatedMsg = "";

		if (toCSVConfig.isWriteFileName()) {
			translatedMsg = (stringFileName + "_" + id2 + "_" + id3 + "_" + target);
		} else if (toCSVConfig.isBeta()){
			translatedMsg = target;
		} else {
			translatedMsg = target;
			if(isFuzzy() || target.contains("-G-")){
				translatedMsg = source;
			}
		}

		return "\""+id+"\",\""+(toCSVConfig.isWriteSource()?source:"")+"\",\""+translatedMsg+"\"\n";

	}

	public StringBuilder toPO() {
		StringBuilder sb = new StringBuilder("\n\n#: ").append(getId());
		if(isFuzzy()) sb.append("\n#, fuzzy");
		sb
				.append("\nmsgctxt \"").append(getId()).append("\"")
				.append("\nmsgid \"").append(getSource()).append("\"")
				.append("\nmsgstr \"\"");
		return sb;
	}

	public StringBuilder toPOT() {
		StringBuilder sb = new StringBuilder("\n\n#: ").append(getId());
		sb
				.append("\nmsgctxt \"").append(getId()).append("\"")
				.append("\nmsgid \"").append(getSource()).append("\"")
				.append("\nmsgstr \"\"");
		return sb;
	}

	public StringBuilder toTranslatedPO() {
		StringBuilder sb = new StringBuilder("\n\n#: ").append(getId());
		if (isFuzzy()) sb.append("\n#, fuzzy");
		sb
				.append("\nmsgctxt \"").append(getId()).append("\"")
				.append("\nmsgid \"").append(getSource()).append("\"")
				.append("\nmsgstr \"").append(getTarget()).append("\"");
		return sb;
	}

	@Override
	public int compareTo(PO o) {
		String src = Integer.toString(o.id2) + o.id3;
		String trg = Integer.toString(this.id2) + this.id3;
		if (src.equals(trg))
			return this.id1.compareTo(o.id1);
		else
			return src.compareTo(trg);
	}
}

File: library/src/main/kotlin/org/metalscraps/eso/lang/lib/bean/PO.kt
package org.metalscraps.eso.lang.lib.bean

import java.util.*


/**
 * Created by 안병길 on 2018-01-18.
 * Whya5448@gmail.com
 */

open class PO(
        val id1: Int,
        val id2: Int,
        val id3: Int,
        private var source: String,
        private var target: String,
        var isFuzzy:Boolean = false,
        var fileName:String = "Undefined"
) : Comparable<PO> {


    fun getID(): String { return "$id1-$id2-$id3" }
    fun getLengthForLang(writeFileName:Boolean = false, beta:Boolean = false) : Int { return getTextForLang(writeFileName, beta).size}
    fun getTextForLang(writeFileName:Boolean = false, beta:Boolean = false): ByteArray { return getText(writeFileName, beta).toByteArray() }
    fun getText(writeFileName:Boolean = false, beta:Boolean = false): String {
        return when {
            writeFileName -> "${fileName}_${id3}_$target"
            beta -> target
            isFuzzy -> source
            else -> target
        }
    }

    fun toPOTFormat(): String {
        return "#: ${getID()}\n" +
                "msgctxt $q${getID()}$q\n" +
                "msgid $q$source$q\n" +
                "msgstr $q$q\n\n"
    }


    override fun compareTo(other: PO): Int {
        val src = Integer.toString(other.id2) + other.id3
        val trg = Integer.toString(this.id2) + this.id3
        return if (src == trg) this.id1.compareTo(other.id1)
        else src.compareTo(trg)
    }

    override fun toString(): String { return "$fileName-$id3,'$source','$target'" }

    companion object {
        private const val q = '"'
        var comparator = Comparator<PO> { o1, o2 ->
            o1.run {
                when {
                    id1 != o2.id1 -> id1.compareTo(o2.id1)
                    id2 != o2.id2 -> id2.compareTo(o2.id2)
                    else -> id3.compareTo(o2.id3)
                }
            }
        }
    }
}